[BUG] ORC Write failing for table which has columns with offset

**Describe the bug**
ORC write fails when input table has column with offset

**Steps/Code to reproduce bug**
```
TYPED_TEST(OrcWriterNumericTypeTest, SingleColumn) {
  auto sequence = cudf::test::make_counting_transform_iterator(
      0, [](auto i) { return TypeParam(i); });
  auto validity = cudf::test::make_counting_transform_iterator(
      0, [](auto i) { return true; });

  constexpr auto num_rows = 100;
  column_wrapper<TypeParam> col(sequence, sequence + num_rows, validity);

  std::vector<std::unique_ptr<column>> cols;
  std::vector<cudf::size_type> ind = {10,50};
  cols.push_back(col.release());
  auto main_table = std::make_unique<table>(std::move(cols));
  auto expected = cudf::experimental::slice(main_table->view(), ind);
  EXPECT_EQ(1, expected[0].num_columns());

  auto filepath = temp_env->get_temp_filepath("OrcSingleColumn.orc");
  cudf_io::write_orc_args out_args{cudf_io::sink_info{filepath},
                                   expected[0]};
  cudf_io::write_orc(out_args);

  cudf_io::read_orc_args in_args{cudf_io::source_info{filepath}};
  in_args.use_index = false;
  auto result = cudf_io::read_orc(in_args);

  expect_tables_equal(expected[0], result.tbl->view());
}

```
Fails with error 
```
E   RuntimeError: CUDA error encountered at: /home/rgsl888/Projects/cudf/cpp/src/io/orc/writer_impl.cu:631: 716 cudaErrorMisalignedAddress misaligned address
```
**Expected behavior**
Should pass the test

**Environment overview (please complete the following information)**
 - Environment location: Bare-metal
 - Method of cuDF install: source

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -117,6 +117,7 @@
 - PR #4251 Add class to docs in `dask-cudf` `derived_from`
 - PR #4261 libxx Cython reorganization
 - PR #4282 Porting nvstrings conversion functions from new libcudf++ to Python/Cython
+- PR #4301 Add support for writing large ORC files in a chunked manner
 
 ## Bug Fixes
 

--- a/cpp/include/cudf/io/functions.hpp
+++ b/cpp/include/cudf/io/functions.hpp
@@ -338,6 +338,88 @@ void write_orc(write_orc_args const& args, rmm::mr::device_memory_resource* mr =
                                                rmm::mr::get_default_resource());
 
 /**
+ * @brief Settings to use for `write_orc_chunked()`
+ */
+struct write_orc_chunked_args {
+  /// Specify the sink to use for writer output
+  sink_info sink;
+  /// Specify the compression format to use
+  compression_type compression;
+  /// Enable writing column statistics
+  bool enable_statistics;
+  /// Optional associated metadata
+  const table_metadata_with_nullability *metadata;
+
+  explicit write_orc_chunked_args(sink_info const& sink_,
+                          const table_metadata_with_nullability *metadata_ = nullptr,
+                          compression_type compression_ = compression_type::AUTO,
+                          bool stats_en = true)
+      : sink(sink_), metadata(metadata_), compression(compression_),
+        enable_statistics(stats_en) {}
+};
+
+/**
+ * @brief Forward declaration of anonymous chunked-writer state struct.
+ */
+namespace detail {
+namespace orc {
+  struct orc_chunked_state;
+};
+};
+
+/**
+ * @brief Begin the process of writing an ORC file in a chunked/stream form.
+ *
+ * The intent of the write_orc_chunked_ path is to allow writing of an
+ * arbitrarily large / arbitrary number of rows to an ORC file in multiple passes.
+ *
+ * The following code snippet demonstrates how to write a single ORC file containing
+ * one logical table by writing a series of individual cudf::tables.
+ * @code
+ *  #include <cudf.h>
+ *  ...
+ *  std::string filepath = "dataset.orc";
+ *  cudf::experimental::io::write_orc_chunked_args args{cudf::sink_info(filepath), table->view()};
+ *  ...
+ *  auto state = cudf::write_orc_chunked_begin(args);
+ *    cudf::write_orc_chunked(table0, state);
+ *    cudf::write_orc_chunked(table1, state);
+ *    ...
+ *  cudf_write_orc_chunked_end(state);
+ * @endcode
+ *
+ * @param[in] args Settings for controlling writing behavior
+ * @param[in] mr Optional resource to use for device memory allocation
+ *
+ * @returns pointer to an anonymous state structure storing information about the chunked write. this
+ *          pointer must be passed to all subsequent write_orc_chunked() and write_orc_chunked_end()
+ *          calls.
+ */
+std::shared_ptr<detail::orc::orc_chunked_state> write_orc_chunked_begin(
+                                                write_orc_chunked_args const& args,
+                                                rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource());
+/**
+ * @brief Write a single table as a subtable of a larger logical orc file/table.
+ *
+ * All tables passed into multiple calls of this function must contain the same # of columns and have columns
+ * of the same type.
+ *
+ * @param[in] table The table data to be written.
+ * @param[in] state Opaque state information about the writer process. Must be the same pointer returned
+ *            from write_orc_chunked_begin()
+ */
+void write_orc_chunked(table_view const& table, std::shared_ptr<detail::orc::orc_chunked_state> state);
+
+/**
+ * @brief Finish writing a chunked/stream orc file.
+ *
+ * @param[in] state Opaque state information about the writer process. Must be the same pointer returned
+ *            from write_orc_chunked_begin()
+ */
+void write_orc_chunked_end(std::shared_ptr<detail::orc::orc_chunked_state>& state);
+
+
+/**
  * @brief Settings to use for `read_parquet()`
  */
 struct read_parquet_args {

--- a/cpp/include/cudf/io/writers.hpp
+++ b/cpp/include/cudf/io/writers.hpp
@@ -26,7 +26,6 @@
 #include <cudf/table/table_view.hpp>
 #include <cudf/types.hpp>
 #include <io/parquet/parquet.h>
-#include <io/orc/orc.h>
 
 #include <memory>
 #include <utility>
@@ -141,32 +140,6 @@ class writer {
    */
   void write_chunked_end(struct orc_chunked_state& state);
 };
-
-
-/**
- * @brief Chunked writer state struct. Contains various pieces of information
- *        needed that span the begin() / write() / end() call process.
- */
-struct orc_chunked_state {
-  /// The writer to be used
-  std::unique_ptr<writer>             wp;
-  /// Cuda stream to be used
-  cudaStream_t                        stream;
-  /// Overall file metadata.  Filled in during the process and written during write_chunked_end()
-  cudf::io::orc::FileFooter           ff;
-  cudf::io::orc::Metadata             md;
-  /// current write position for rowgroups/chunks
-  size_t                              current_chunk_offset;
-  /// optional user metadata
-  table_metadata const*               user_metadata = nullptr;
-  /// only used in the write_chunked() case. copied from the (optionally) user supplied
-  /// argument to write_chunked_begin()
-  table_metadata_with_nullability     user_metadata_with_nullability;
-  /// special parameter only used by detail::write() to indicate that we are guaranteeing
-  /// a single table write.  this enables some internal optimizations.
-  bool                                single_write_mode = false;
-};
-
 
 }  // namespace orc
 

--- a/cpp/include/cudf/io/writers.hpp
+++ b/cpp/include/cudf/io/writers.hpp
@@ -26,6 +26,7 @@
 #include <cudf/table/table_view.hpp>
 #include <cudf/types.hpp>
 #include <io/parquet/parquet.h>
+#include <io/orc/orc.h>
 
 #include <memory>
 #include <utility>
@@ -117,7 +118,54 @@ class writer {
    * @param stream Optional stream to use for device memory alloc and kernels
    */
   void write_all(table_view const& table, const table_metadata *metadata = nullptr, cudaStream_t stream = 0);
+
+  /**
+   * @brief Begins the chunked/streamed write process.
+   *
+   * @param[in] state State information that crosses _begin() / write_chunked() / _end() boundaries.
+   */
+  void write_chunked_begin(struct orc_chunked_state& state);
+
+  /**
+   * @brief Writes a single subtable as part of a larger ORC file/table write.
+   *
+   * @param[in] table The table information to be written
+   * @param[in] state State information that crosses _begin() / write_chunked() / _end() boundaries.
+   */
+  void write_chunked(table_view const& table, struct orc_chunked_state& state);
+
+  /**
+   * @brief Finishes the chunked/streamed write process.
+   *
+   * @param[in] state State information that crosses _begin() / write_chunked() / _end() boundaries.
+   */
+  void write_chunked_end(struct orc_chunked_state& state);
 };
+
+
+/**
+ * @brief Chunked writer state struct. Contains various pieces of information
+ *        needed that span the begin() / write() / end() call process.
+ */
+struct orc_chunked_state {
+  /// The writer to be used
+  std::unique_ptr<writer>             wp;
+  /// Cuda stream to be used
+  cudaStream_t                        stream;
+  /// Overall file metadata.  Filled in during the process and written during write_chunked_end()
+  cudf::io::orc::FileFooter           ff;
+  /// current write position for rowgroups/chunks
+  size_t                              current_chunk_offset;
+  /// optional user metadata
+  table_metadata const*               user_metadata = nullptr;
+  /// only used in the write_chunked() case. copied from the (optionally) user supplied
+  /// argument to write_chunked_begin()
+  table_metadata_with_nullability     user_metadata_with_nullability;
+  /// special parameter only used by detail::write() to indicate that we are guaranteeing
+  /// a single table write.  this enables some internal optimizations.
+  bool                                single_write_mode = false;
+};
+
 
 }  // namespace orc
 

--- a/cpp/include/cudf/io/writers.hpp
+++ b/cpp/include/cudf/io/writers.hpp
@@ -154,6 +154,7 @@ struct orc_chunked_state {
   cudaStream_t                        stream;
   /// Overall file metadata.  Filled in during the process and written during write_chunked_end()
   cudf::io::orc::FileFooter           ff;
+  cudf::io::orc::Metadata             md;
   /// current write position for rowgroups/chunks
   size_t                              current_chunk_offset;
   /// optional user metadata

--- a/cpp/src/io/functions.cpp
+++ b/cpp/src/io/functions.cpp
@@ -143,11 +143,11 @@ table_with_metadata read_csv(read_csv_args const& args,
   }
 }
 
+namespace orc = cudf::experimental::io::detail::orc;
+
 // Freeform API wraps the detail reader class API
 table_with_metadata read_orc(read_orc_args const& args,
                                 rmm::mr::device_memory_resource* mr) {
-  namespace orc = cudf::experimental::io::detail::orc;
-
   orc::reader_options options{args.columns, args.use_index, args.use_np_dtypes,
                               args.timestamp_type, args.decimals_as_float,
                               args.forced_decimals_scale};
@@ -165,13 +165,50 @@ table_with_metadata read_orc(read_orc_args const& args,
 // Freeform API wraps the detail writer class API
 void write_orc(write_orc_args const& args,
                rmm::mr::device_memory_resource* mr) {
-  namespace orc = cudf::experimental::io::detail::orc;
-
   orc::writer_options options{args.compression, args.enable_statistics};
   auto writer = make_writer<orc::writer>(args.sink, options, mr);
 
   writer->write_all(args.table, args.metadata);
 }
+
+/**
+ * @copydoc cudf::experimental::io::write_orc_chunked_begin
+ *
+ **/
+std::shared_ptr<orc::orc_chunked_state> write_orc_chunked_begin(write_orc_chunked_args const& args, rmm::mr::device_memory_resource* mr){
+  orc::writer_options options{args.compression, args.enable_statistics};
+
+  auto state = std::make_shared<orc::orc_chunked_state>();
+  state->wp = make_writer<orc::writer>(args.sink, options, mr);
+
+  // have to make a copy of the metadata here since we can't really
+  // guarantee the lifetime of the incoming pointer
+  if(args.metadata != nullptr){
+    state->user_metadata_with_nullability = *args.metadata;
+    state->user_metadata = &state->user_metadata_with_nullability;
+  }
+  state->stream = 0;
+  state->wp->write_chunked_begin(*state);
+  return state;
+}
+
+/**
+ * @copydoc cudf::experimental::io::write_orc_chunked
+ *
+ **/
+void write_orc_chunked(table_view const& table, std::shared_ptr<orc::orc_chunked_state> state){
+  state->wp->write_chunked(table, *state);
+}
+
+/**
+ * @copydoc cudf::experimental::io::write_orc_chunked_end
+ *
+ **/
+void write_orc_chunked_end(std::shared_ptr<orc::orc_chunked_state>& state){
+  state->wp->write_chunked_end(*state);
+  state.reset();
+}
+
 
 // Freeform API wraps the detail reader class API
 table_with_metadata read_parquet(read_parquet_args const& args,

--- a/cpp/src/io/functions.cpp
+++ b/cpp/src/io/functions.cpp
@@ -19,7 +19,7 @@
 #include <cudf/io/writers.hpp>
 #include <cudf/table/table.hpp>
 #include <cudf/utilities/error.hpp>
-
+#include "orc/writer_impl.hpp" // For orc_chunked_state definition
 
 namespace cudf {
 namespace experimental {

--- a/cpp/src/io/functions.cpp
+++ b/cpp/src/io/functions.cpp
@@ -19,7 +19,7 @@
 #include <cudf/io/writers.hpp>
 #include <cudf/table/table.hpp>
 #include <cudf/utilities/error.hpp>
-#include "orc/writer_impl.hpp" // For orc_chunked_state definition
+#include "orc/chunked_state.hpp" // orc_chunked_state definition
 
 namespace cudf {
 namespace experimental {

--- a/cpp/src/io/orc/chunked_state.hpp
+++ b/cpp/src/io/orc/chunked_state.hpp
@@ -1,0 +1,70 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @file chunked_state.hpp
+ * @brief definition for chunked state structure used by ORC writer
+ */
+
+#pragma once
+
+#include "orc.h"
+
+#include <cudf/io/writers.hpp>
+#include <cudf/table/table.hpp>
+#include <cudf/utilities/error.hpp>
+#include <io/utilities/data_sink.hpp>
+
+#include <memory>
+#include <string>
+#include <vector>
+
+namespace cudf {
+namespace experimental {
+namespace io {
+namespace detail {
+namespace orc {
+
+/**
+ * @brief Chunked writer state struct. Contains various pieces of information
+ *        needed that span the begin() / write() / end() call process.
+ */
+struct orc_chunked_state {
+  /// The writer to be used
+  std::unique_ptr<writer>             wp;
+  /// Cuda stream to be used
+  cudaStream_t                        stream;
+  /// Overall file metadata.  Filled in during the process and written during write_chunked_end()
+  cudf::io::orc::FileFooter           ff;
+  cudf::io::orc::Metadata             md;
+  /// current write position for rowgroups/chunks
+  size_t                              current_chunk_offset;
+  /// optional user metadata
+  table_metadata const*               user_metadata = nullptr;
+  /// only used in the write_chunked() case. copied from the (optionally) user supplied
+  /// argument to write_chunked_begin()
+  table_metadata_with_nullability     user_metadata_with_nullability;
+  /// special parameter only used by detail::write() to indicate that we are guaranteeing
+  /// a single table write.  this enables some internal optimizations.
+  bool                                single_write_mode = false;
+};
+
+
+}  // namespace orc
+}  // namespace detail
+}  // namespace io
+}  // namespace experimental
+}  // namespace cudf

--- a/cpp/src/io/orc/orc.h
+++ b/cpp/src/io/orc/orc.h
@@ -150,6 +150,7 @@ public:
     DECL_ORC_STRUCT(ColumnEncoding);
     DECL_ORC_STRUCT(StripeStatistics);
     DECL_ORC_STRUCT(Metadata);
+#undef DECL_ORC_STRUCT
 protected:
     bool InitSchema(FileFooter *);
 
@@ -185,6 +186,7 @@ public:
     DECL_PBW_STRUCT(ColumnEncoding);
     DECL_PBW_STRUCT(StripeStatistics);
     DECL_PBW_STRUCT(Metadata);
+#undef DECL_PBW_STRUCT
 protected:
     std::vector<uint8_t> *m_buf;
 };

--- a/cpp/src/io/orc/writer_impl.cu
+++ b/cpp/src/io/orc/writer_impl.cu
@@ -901,16 +901,14 @@ void writer::impl::write_data_stream(gpu::StripeStream const &strm_desc,
 void writer::impl::add_uncompressed_block_headers(std::vector<uint8_t>& v) {
   if (compression_kind_ != NONE) {
     size_t uncomp_len = v.size() - 3, pos = 0, block_len;
-    if (uncomp_len > compression_blocksize_) {
+    while (uncomp_len > compression_blocksize_) {
       block_len = compression_blocksize_ * 2 + 1;
-      while (uncomp_len > compression_blocksize_) {
-        v[pos + 0] = static_cast<uint8_t>(block_len >> 0);
-        v[pos + 1] = static_cast<uint8_t>(block_len >> 8);
-        v[pos + 2] = static_cast<uint8_t>(block_len >> 16);
-        pos += 3 + compression_blocksize_;
-        v.insert(v.begin() + pos, 3, 0);
-        uncomp_len -= compression_blocksize_;
-      }
+      v[pos + 0] = static_cast<uint8_t>(block_len >> 0);
+      v[pos + 1] = static_cast<uint8_t>(block_len >> 8);
+      v[pos + 2] = static_cast<uint8_t>(block_len >> 16);
+      pos += 3 + compression_blocksize_;
+      v.insert(v.begin() + pos, 3, 0);
+      uncomp_len -= compression_blocksize_;
     }
     block_len = uncomp_len * 2 + 1;
     v[pos + 0] = static_cast<uint8_t>(block_len >> 0);

--- a/cpp/src/io/orc/writer_impl.cu
+++ b/cpp/src/io/orc/writer_impl.cu
@@ -1189,6 +1189,15 @@ void writer::impl::write_chunked(table_view const& table, orc_chunked_state& sta
       state.ff.types[0].fieldNames[i] = orc_columns[i].orc_name();
     }
   }
+  else {
+    // verify the user isn't passing mismatched tables
+    CUDF_EXPECTS(state.ff.types.size() == 1 + orc_columns.size(),
+                 "Mismatch in table structure between multiple calls to write_chunked");    
+    for (auto i = 0; i < num_columns; i++) {
+      CUDF_EXPECTS(state.ff.types[1 + i].kind == orc_columns[i].orc_kind(),
+                   "Mismatch in column types between multiple calls to write_chunked");
+    }
+  }
   state.ff.stripes.insert(state.ff.stripes.end(),
                           std::make_move_iterator(stripes.begin()),
                           std::make_move_iterator(stripes.end()));

--- a/cpp/src/io/orc/writer_impl.cu
+++ b/cpp/src/io/orc/writer_impl.cu
@@ -898,6 +898,27 @@ void writer::impl::write_data_stream(gpu::StripeStream const &strm_desc,
   stripe.dataLength += length;
 }
 
+void writer::impl::add_uncompressed_block_headers(std::vector<uint8_t>& v) {
+  if (compression_kind_ != NONE) {
+    size_t uncomp_len = v.size() - 3, pos = 0, block_len;
+    if (uncomp_len > compression_blocksize_) {
+      block_len = compression_blocksize_ * 2 + 1;
+      while (uncomp_len > compression_blocksize_) {
+        v[pos + 0] = static_cast<uint8_t>(block_len >> 0);
+        v[pos + 1] = static_cast<uint8_t>(block_len >> 8);
+        v[pos + 2] = static_cast<uint8_t>(block_len >> 16);
+        pos += 3 + compression_blocksize_;
+        v.insert(v.begin() + pos, 3, 0);
+        uncomp_len -= compression_blocksize_;
+      }
+    }
+    block_len = uncomp_len * 2 + 1;
+    v[pos + 0] = static_cast<uint8_t>(block_len >> 0);
+    v[pos + 1] = static_cast<uint8_t>(block_len >> 8);
+    v[pos + 2] = static_cast<uint8_t>(block_len >> 16);
+  }
+}
+
 writer::impl::impl(std::unique_ptr<data_sink> sink, writer_options const &options,
   rmm::mr::device_memory_resource *mr):
   compression_kind_(to_orc_compression(options.compression)), 
@@ -1218,13 +1239,8 @@ void writer::impl::write_chunked_end(orc_chunked_state& state) {
   if (state.md.stripeStats.size() != 0) {
     buffer_.resize((compression_kind_ != NONE) ? 3 : 0);
     pbw_.write(&state.md);
+    add_uncompressed_block_headers(buffer_);
     ps.metadataLength = buffer_.size();
-    if (compression_kind_ != NONE) {
-      uint32_t uncomp_md_len = (uint32_t)(ps.metadataLength - 3) * 2 + 1;
-      buffer_[0] = static_cast<uint8_t>(uncomp_md_len >> 0);
-      buffer_[1] = static_cast<uint8_t>(uncomp_md_len >> 8);
-      buffer_[2] = static_cast<uint8_t>(uncomp_md_len >> 16);
-    }
     out_sink_->write(buffer_.data(), buffer_.size());
   }
   else {
@@ -1232,6 +1248,7 @@ void writer::impl::write_chunked_end(orc_chunked_state& state) {
   }
   buffer_.resize((compression_kind_ != NONE) ? 3 : 0);
   pbw_.write(&state.ff);
+  add_uncompressed_block_headers(buffer_);
 
   // Write postscript metadata
   ps.footerLength = buffer_.size();
@@ -1239,14 +1256,6 @@ void writer::impl::write_chunked_end(orc_chunked_state& state) {
   ps.compressionBlockSize = compression_blocksize_;
   ps.version = {0, 12};
   ps.magic = MAGIC;
-  if (compression_kind_ != NONE) {
-    // TODO: If the file footer ends up larger than the compression block
-    // size, we'll need to insert additional 3-byte block headers
-    uint32_t uncomp_ff_len = (uint32_t)(ps.footerLength - 3) * 2 + 1;
-    buffer_[0] = static_cast<uint8_t>(uncomp_ff_len >> 0);
-    buffer_[1] = static_cast<uint8_t>(uncomp_ff_len >> 8);
-    buffer_[2] = static_cast<uint8_t>(uncomp_ff_len >> 16);
-  }
   const auto ps_length = static_cast<uint8_t>(pbw_.write(&ps));
   buffer_.push_back(ps_length);
   out_sink_->write(buffer_.data(), buffer_.size());

--- a/cpp/src/io/orc/writer_impl.cu
+++ b/cpp/src/io/orc/writer_impl.cu
@@ -356,7 +356,8 @@ void writer::impl::build_dictionaries(
 
 std::vector<Stream> writer::impl::gather_streams(
     orc_column_view *columns, size_t num_columns, size_t num_rows,
-    std::vector<uint32_t> const &stripe_list, std::vector<int32_t> &strm_ids) {
+    std::vector<uint32_t> const &stripe_list, std::vector<int32_t> &strm_ids,
+    const orc_chunked_state& state) {
   // First n + 1 streams are row index streams, including 'column 0'
   std::vector<Stream> streams;
   streams.resize(num_columns + 1);
@@ -374,7 +375,15 @@ std::vector<Stream> writer::impl::gather_streams(
     int64_t data_stream_size = 0;
     int64_t data2_stream_size = 0;
     int64_t dict_stream_size = 0;
-    if (columns[i].nullable() || columns[i].data_count() != num_rows) {
+    bool is_nullable;
+
+    if (state.single_write_mode){
+      is_nullable = (columns[i].nullable() || columns[i].data_count() < num_rows);
+    } else {
+      is_nullable = (i < state.user_metadata_with_nullability.column_nullable.size()) ?
+                              state.user_metadata_with_nullability.column_nullable[i] : true;
+    }
+    if (is_nullable) {
       present_stream_size = ((row_index_stride_ + 7) >> 3);
       present_stream_size += (present_stream_size + 0x7f) >> 7;
     }
@@ -919,6 +928,11 @@ void writer::impl::write_chunked(table_view const& table, orc_chunked_state& sta
   // Mapping of string columns for quick look-up
   std::vector<int> str_col_ids;
 
+  if(state.user_metadata_with_nullability.column_nullable.size() > 0){
+    CUDF_EXPECTS(state.user_metadata_with_nullability.column_nullable.size() == static_cast<size_t>(num_columns),
+                 "When passing values in user_metadata_with_nullability, data for all columns must be specified");
+  }
+
   // Wrapper around cudf columns to attach ORC-specific type info
   std::vector<orc_column_view> orc_columns;
   orc_columns.reserve(num_columns); // Avoids unnecessary re-allocation
@@ -988,7 +1002,7 @@ void writer::impl::write_chunked(table_view const& table, orc_chunked_state& sta
   // Initialize streams
   std::vector<int32_t> strm_ids(num_columns * gpu::CI_NUM_STREAMS, -1);
   auto streams = gather_streams(orc_columns.data(), orc_columns.size(),
-                                num_rows, stripe_list, strm_ids);
+                                num_rows, stripe_list, strm_ids, state);
 
   // Encode column data chunks
   const auto num_chunks = num_rowgroups * num_columns;

--- a/cpp/src/io/orc/writer_impl.cu
+++ b/cpp/src/io/orc/writer_impl.cu
@@ -164,7 +164,7 @@ class orc_column_view {
         _type_width(_string_type ? 0 : cudf::size_of(col.type())),
         _data_count(col.size()),
         _null_count(col.null_count()),
-        _data(col.data<uint8_t>()),
+        _data(col.head<uint8_t>() + col.offset() * _type_width),
         _nulls(col.nullable() ? col.null_mask() : nullptr),
         _clockscale(to_clockscale<uint8_t>(col.type().id())),
         _type_kind(to_orc_type(col.type().id())) {

--- a/cpp/src/io/orc/writer_impl.cu
+++ b/cpp/src/io/orc/writer_impl.cu
@@ -897,6 +897,22 @@ writer::impl::impl(std::unique_ptr<data_sink> sink, writer_options const &option
   _mr(mr) {}
 
 void writer::impl::write(table_view const &table, const table_metadata *metadata, cudaStream_t stream) {
+  orc_chunked_state state;
+  state.user_metadata = metadata;
+  state.stream = stream;
+  state.single_write_mode = true;
+
+  write_chunked_begin(state);
+  write_chunked(table, state);
+  write_chunked_end(state);    
+}
+
+void writer::impl::write_chunked_begin(orc_chunked_state& state) {
+  // Write file header
+  out_sink_->write(MAGIC, std::strlen(MAGIC));
+}
+
+void writer::impl::write_chunked(table_view const& table, orc_chunked_state& state) {
   size_type num_columns = table.num_columns();
   size_type num_rows = 0;
 
@@ -912,7 +928,7 @@ void writer::impl::write(table_view const &table, const table_metadata *metadata
     const auto current_str_id = str_col_ids.size();
 
     num_rows = std::max<uint32_t>(num_rows, col.size());
-    orc_columns.emplace_back(current_id, current_str_id, col, metadata, stream);
+    orc_columns.emplace_back(current_id, current_str_id, col, state.user_metadata, state.stream);
     if (orc_columns.back().is_string()) {
       str_col_ids.push_back(current_id);
     }
@@ -928,7 +944,7 @@ void writer::impl::write(table_view const &table, const table_metadata *metadata
   if (str_col_ids.size() != 0) {
     init_dictionaries(orc_columns.data(), num_rows, str_col_ids,
                       dict_data.data().get(), dict_index.data().get(), dict,
-                      stream);
+                      state.stream);
   }
 
   // Decide stripe boundaries early on, based on uncompressed size
@@ -966,7 +982,7 @@ void writer::impl::write(table_view const &table, const table_metadata *metadata
   hostdevice_vector<gpu::StripeDictionary> stripe_dict(num_stripe_dict);
   if (str_col_ids.size() != 0) {
     build_dictionaries(orc_columns.data(), num_rows, str_col_ids, stripe_list,
-                       dict, dict_index.data().get(), stripe_dict, stream);
+                       dict, dict_index.data().get(), stripe_dict, state.stream);
   }
 
   // Initialize streams
@@ -979,7 +995,7 @@ void writer::impl::write(table_view const &table, const table_metadata *metadata
   hostdevice_vector<gpu::EncChunk> chunks(num_chunks);
   auto output = encode_columns(orc_columns.data(), num_columns, num_rows,
                                num_rowgroups, str_col_ids, stripe_list, streams,
-                               strm_ids, chunks, stream);
+                               strm_ids, chunks, state.stream);
 
   // Assemble individual desparate column chunks into contiguous data streams
   const auto num_index_streams = (num_columns + 1);
@@ -988,13 +1004,13 @@ void writer::impl::write(table_view const &table, const table_metadata *metadata
   hostdevice_vector<gpu::StripeStream> strm_desc(num_stripe_streams);
   auto stripes =
       gather_stripes(num_columns, num_rows, num_index_streams, num_data_streams,
-                     stripe_list, chunks, strm_desc, stream);
+                     stripe_list, chunks, strm_desc, state.stream);
 
   // Gather column statistics
   std::vector<std::vector<uint8_t>> column_stats;
   if (enable_statistics_ && num_columns > 0 && num_rows > 0) {
     column_stats = gather_statistic_blobs(orc_columns.data(), num_columns, num_rows,
-                                          num_rowgroups, stripe_list, stripes, chunks, stream);
+                                          num_rowgroups, stripe_list, stripes, chunks, state.stream);
   }
 
   // Allocate intermediate output stream buffer
@@ -1032,31 +1048,28 @@ void writer::impl::write(table_view const &table, const table_metadata *metadata
   }();
 
   // Compress the data streams
-  rmm::device_buffer compressed_data(compressed_bfr_size, stream);
+  rmm::device_buffer compressed_data(compressed_bfr_size, state.stream);
   hostdevice_vector<gpu_inflate_status_s> comp_out(num_compressed_blocks);
   hostdevice_vector<gpu_inflate_input_s> comp_in(num_compressed_blocks);
   if (compression_kind_ != NONE) {
     CUDA_TRY(cudaMemcpyAsync(strm_desc.device_ptr(), strm_desc.host_ptr(),
                              strm_desc.memory_size(), cudaMemcpyHostToDevice,
-                             stream));
+                             state.stream));
     CUDA_TRY(gpu::CompressOrcDataStreams(
         static_cast<uint8_t *>(compressed_data.data()), strm_desc.device_ptr(),
         chunks.device_ptr(), comp_in.device_ptr(), comp_out.device_ptr(),
         num_stripe_streams, num_compressed_blocks, compression_kind_,
-        compression_blocksize_, stream));
+        compression_blocksize_, state.stream));
     CUDA_TRY(cudaMemcpyAsync(strm_desc.host_ptr(), strm_desc.device_ptr(),
                              strm_desc.memory_size(), cudaMemcpyDeviceToHost,
-                             stream));
+                             state.stream));
     CUDA_TRY(cudaMemcpyAsync(comp_out.host_ptr(), comp_out.device_ptr(),
                              comp_out.memory_size(), cudaMemcpyDeviceToHost,
-                             stream));
-    CUDA_TRY(cudaStreamSynchronize(stream));
+                             state.stream));
+    CUDA_TRY(cudaStreamSynchronize(state.stream));
   }
 
   ProtobufWriter pbw_(&buffer_);
-
-  // Write file header
-  out_sink_->write(MAGIC, std::strlen(MAGIC));
 
   // Write stripes
   size_t group = 0;
@@ -1081,7 +1094,7 @@ void writer::impl::write(table_view const &table, const table_metadata *metadata
 
       write_data_stream(ss, ck, static_cast<uint8_t *>(compressed_data.data()),
                         stream_output.get(), stripes[stripe_id], streams,
-                        stream);
+                        state.stream);
     }
 
     // Write stripefooter consisting of stream information
@@ -1114,64 +1127,74 @@ void writer::impl::write(table_view const &table, const table_metadata *metadata
     group += groups_in_stripe;
   }
 
-  // Write filefooter metadata
-  Metadata md;
-  FileFooter ff;
-  PostScript ps;
-
   if (column_stats.size() != 0) {
     // File-level statistics
-    ff.statistics.resize(1 + num_columns);
-    // First entry contains total number of rows
-    buffer_.resize(0);
-    pbw_.putb(1 * 8 + PB_TYPE_VARINT);
-    pbw_.put_uint(num_rows);
-    ff.statistics[0] = std::move(buffer_);
-    for (int i = 0; i < num_columns; i++) {
-      size_t idx = stripe_list.size() * num_columns + i;
-      if (idx < column_stats.size()) {
-        ff.statistics[1 + i] = std::move(column_stats[idx]);
+    // NOTE: Excluded from chunked write mode to avoid the need for merging stats accross calls
+    if (state.single_write_mode) {
+      state.ff.statistics.resize(1 + num_columns);
+      // First entry contains total number of rows
+      buffer_.resize(0);
+      pbw_.putb(1 * 8 + PB_TYPE_VARINT);
+      pbw_.put_uint(num_rows);
+      state.ff.statistics[0] = std::move(buffer_);
+      for (int i = 0; i < num_columns; i++) {
+        size_t idx = stripe_list.size() * num_columns + i;
+        if (idx < column_stats.size()) {
+          state.ff.statistics[1 + i] = std::move(column_stats[idx]);
+        }
       }
     }
     // Stripe-level statistics
-    md.stripeStats.resize(stripe_list.size());
+    size_t first_stripe = state.md.stripeStats.size();
+    state.md.stripeStats.resize(first_stripe + stripe_list.size());
     for (size_t stripe_id = 0; stripe_id < stripe_list.size(); stripe_id++) {
-      md.stripeStats[stripe_id].colStats.resize(1 + num_columns);
+      state.md.stripeStats[first_stripe + stripe_id].colStats.resize(1 + num_columns);
       buffer_.resize(0);
       pbw_.putb(1 * 8 + PB_TYPE_VARINT);
       pbw_.put_uint(stripes[stripe_id].numberOfRows);
-      md.stripeStats[stripe_id].colStats[0] = std::move(buffer_);
+      state.md.stripeStats[first_stripe + stripe_id].colStats[0] = std::move(buffer_);
       for (int i = 0; i < num_columns; i++) {
         size_t idx = stripe_list.size() * i + stripe_id;
         if (idx < column_stats.size()) {
-          md.stripeStats[stripe_id].colStats[1 + i] = std::move(column_stats[idx]);
+          state.md.stripeStats[first_stripe + stripe_id].colStats[1 + i] = std::move(column_stats[idx]);
         }
       }
     }
   }
-  ff.headerLength = std::strlen(MAGIC);
-  ff.contentLength = out_sink_->bytes_written();
-  ff.stripes = std::move(stripes);
-  ff.numberOfRows = num_rows;
-  ff.rowIndexStride = row_index_stride_;
-  ff.types.resize(1 + num_columns);
-  ff.types[0].kind = STRUCT;
-  ff.types[0].subtypes.resize(num_columns);
-  ff.types[0].fieldNames.resize(num_columns);
-  for (int i = 0; i < num_columns; ++i) {
-    ff.types[1 + i].kind = orc_columns[i].orc_kind();
-    ff.types[0].subtypes[i] = 1 + i;
-    ff.types[0].fieldNames[i] = orc_columns[i].orc_name();
+  if (state.ff.headerLength == 0) {
+    // First call
+    state.ff.headerLength = std::strlen(MAGIC);
+    state.ff.rowIndexStride = row_index_stride_;
+    state.ff.types.resize(1 + num_columns);
+    state.ff.types[0].kind = STRUCT;
+    state.ff.types[0].subtypes.resize(num_columns);
+    state.ff.types[0].fieldNames.resize(num_columns);
+    for (int i = 0; i < num_columns; ++i) {
+      state.ff.types[1 + i].kind = orc_columns[i].orc_kind();
+      state.ff.types[0].subtypes[i] = 1 + i;
+      state.ff.types[0].fieldNames[i] = orc_columns[i].orc_name();
+    }
   }
-  if (metadata) {
-    for (auto it = metadata->user_data.begin(); it != metadata->user_data.end(); it++) {
-      ff.metadata.push_back({it->first, it->second});
+  state.ff.stripes.insert(state.ff.stripes.end(),
+                          std::make_move_iterator(stripes.begin()),
+                          std::make_move_iterator(stripes.end()));
+  state.ff.numberOfRows += num_rows;
+}
+
+void writer::impl::write_chunked_end(orc_chunked_state& state) {
+  ProtobufWriter pbw_(&buffer_);
+  PostScript ps;
+
+  state.ff.contentLength = out_sink_->bytes_written();
+  if (state.user_metadata) {
+    for (auto it = state.user_metadata->user_data.begin(); it != state.user_metadata->user_data.end(); it++) {
+      state.ff.metadata.push_back({it->first, it->second});
     }
   }
   // Write statistics metadata
-  if (md.stripeStats.size() != 0) {
+  if (state.md.stripeStats.size() != 0) {
     buffer_.resize((compression_kind_ != NONE) ? 3 : 0);
-    pbw_.write(&md);
+    pbw_.write(&state.md);
     ps.metadataLength = buffer_.size();
     if (compression_kind_ != NONE) {
       uint32_t uncomp_md_len = (uint32_t)(ps.metadataLength - 3) * 2 + 1;
@@ -1185,7 +1208,7 @@ void writer::impl::write(table_view const &table, const table_metadata *metadata
     ps.metadataLength = 0;
   }
   buffer_.resize((compression_kind_ != NONE) ? 3 : 0);
-  pbw_.write(&ff);
+  pbw_.write(&state.ff);
 
   // Write postscript metadata
   ps.footerLength = buffer_.size();
@@ -1207,6 +1230,7 @@ void writer::impl::write(table_view const &table, const table_metadata *metadata
   out_sink_->flush();
 }
 
+
 // Forward to implementation
 writer::writer(std::string const& filepath, writer_options const& options,
                rmm::mr::device_memory_resource *mr)
@@ -1226,6 +1250,21 @@ writer::~writer() = default;
 // Forward to implementation
 void writer::write_all(table_view const &table, const table_metadata *metadata, cudaStream_t stream) {
   _impl->write(table, metadata, stream);
+}
+
+// Forward to implementation
+void writer::write_chunked_begin(orc_chunked_state& state){
+  _impl->write_chunked_begin(state);
+}
+
+// Forward to implementation
+void writer::write_chunked(table_view const& table, orc_chunked_state &state){
+  _impl->write_chunked(table, state);
+}
+
+// Forward to implementation
+void writer::write_chunked_end(orc_chunked_state &state){
+  _impl->write_chunked_end(state);
 }
 
 }  // namespace orc

--- a/cpp/src/io/orc/writer_impl.hpp
+++ b/cpp/src/io/orc/writer_impl.hpp
@@ -84,6 +84,28 @@ class writer::impl {
    **/
   void write(table_view const& table, const table_metadata *metadata, cudaStream_t stream);
 
+  /**
+   * @brief Begins the chunked/streamed write process.
+   *
+   * @param[in] orc_chunked_state State information that crosses _begin() / write_chunked() / _end() boundaries.
+   */
+  void write_chunked_begin(orc_chunked_state& state);
+
+  /**
+   * @brief Writes a single subtable as part of a larger parquet file/table write.
+   *
+   * @param[in] table The table information to be written
+   * @param[in] orc_chunked_state State information that crosses _begin() / write_chunked() / _end() boundaries.
+   */
+  void write_chunked(table_view const& table, orc_chunked_state& state);
+
+  /**
+   * @brief Finishes the chunked/streamed write process.
+   *
+   * @param[in] orc_chunked_state State information that crosses _begin() / write_chunked() / _end() boundaries.
+   */
+  void write_chunked_end(orc_chunked_state& state);
+
  private:
   /**
    * @brief Builds up column dictionaries indices

--- a/cpp/src/io/orc/writer_impl.hpp
+++ b/cpp/src/io/orc/writer_impl.hpp
@@ -271,6 +271,13 @@ class writer::impl {
                          std::vector<Stream>& streams, cudaStream_t stream);
 
   /**
+   * @brief Insert 3-byte uncompressed block headers in a byte vector
+   *
+   * @param byte_vector Raw data (must include initial 3-byte header)
+   */
+  void add_uncompressed_block_headers(std::vector<uint8_t>& byte_vector);
+
+  /**
    * @brief Returns the number of row groups for holding the specified rows
    *
    * @tparam T Optional type

--- a/cpp/src/io/orc/writer_impl.hpp
+++ b/cpp/src/io/orc/writer_impl.hpp
@@ -36,36 +36,13 @@
 #include <string>
 #include <vector>
 
+#include "chunked_state.hpp"
+
 namespace cudf {
 namespace experimental {
 namespace io {
 namespace detail {
 namespace orc {
-
-/**
- * @brief Chunked writer state struct. Contains various pieces of information
- *        needed that span the begin() / write() / end() call process.
- */
-struct orc_chunked_state {
-  /// The writer to be used
-  std::unique_ptr<writer>             wp;
-  /// Cuda stream to be used
-  cudaStream_t                        stream;
-  /// Overall file metadata.  Filled in during the process and written during write_chunked_end()
-  cudf::io::orc::FileFooter           ff;
-  cudf::io::orc::Metadata             md;
-  /// current write position for rowgroups/chunks
-  size_t                              current_chunk_offset;
-  /// optional user metadata
-  table_metadata const*               user_metadata = nullptr;
-  /// only used in the write_chunked() case. copied from the (optionally) user supplied
-  /// argument to write_chunked_begin()
-  table_metadata_with_nullability     user_metadata_with_nullability;
-  /// special parameter only used by detail::write() to indicate that we are guaranteeing
-  /// a single table write.  this enables some internal optimizations.
-  bool                                single_write_mode = false;
-};
-
 
 // Forward internal classes
 class orc_column_view;

--- a/cpp/src/io/orc/writer_impl.hpp
+++ b/cpp/src/io/orc/writer_impl.hpp
@@ -158,7 +158,8 @@ class writer::impl {
   std::vector<Stream> gather_streams(orc_column_view* columns,
                                      size_t num_columns, size_t num_rows,
                                      std::vector<uint32_t> const& stripe_list,
-                                     std::vector<int32_t>& strm_ids);
+                                     std::vector<int32_t>& strm_ids,
+                                     const orc_chunked_state& state);
 
   /**
    * @brief Encodes the streams as a series of column data chunks

--- a/cpp/src/io/orc/writer_impl.hpp
+++ b/cpp/src/io/orc/writer_impl.hpp
@@ -92,7 +92,7 @@ class writer::impl {
   void write_chunked_begin(orc_chunked_state& state);
 
   /**
-   * @brief Writes a single subtable as part of a larger parquet file/table write.
+   * @brief Writes a single subtable as part of a larger ORC file/table write.
    *
    * @param[in] table The table information to be written
    * @param[in] orc_chunked_state State information that crosses _begin() / write_chunked() / _end() boundaries.

--- a/cpp/src/io/orc/writer_impl.hpp
+++ b/cpp/src/io/orc/writer_impl.hpp
@@ -42,6 +42,31 @@ namespace io {
 namespace detail {
 namespace orc {
 
+/**
+ * @brief Chunked writer state struct. Contains various pieces of information
+ *        needed that span the begin() / write() / end() call process.
+ */
+struct orc_chunked_state {
+  /// The writer to be used
+  std::unique_ptr<writer>             wp;
+  /// Cuda stream to be used
+  cudaStream_t                        stream;
+  /// Overall file metadata.  Filled in during the process and written during write_chunked_end()
+  cudf::io::orc::FileFooter           ff;
+  cudf::io::orc::Metadata             md;
+  /// current write position for rowgroups/chunks
+  size_t                              current_chunk_offset;
+  /// optional user metadata
+  table_metadata const*               user_metadata = nullptr;
+  /// only used in the write_chunked() case. copied from the (optionally) user supplied
+  /// argument to write_chunked_begin()
+  table_metadata_with_nullability     user_metadata_with_nullability;
+  /// special parameter only used by detail::write() to indicate that we are guaranteeing
+  /// a single table write.  this enables some internal optimizations.
+  bool                                single_write_mode = false;
+};
+
+
 // Forward internal classes
 class orc_column_view;
 

--- a/cpp/src/io/parquet/writer_impl.cu
+++ b/cpp/src/io/parquet/writer_impl.cu
@@ -114,7 +114,7 @@ class parquet_column_view {
         _ts_scale(0),
         _data_count(col.size()),
         _null_count(col.null_count()),
-        _data(col.data<uint8_t>()),
+        _data(col.head<uint8_t>() + col.offset() * _type_width),
         _nulls(col.nullable() ? col.null_mask() : nullptr) {
     switch(col.type().id()) {
       case cudf::type_id::INT8:

--- a/cpp/tests/io/orc_test.cu
+++ b/cpp/tests/io/orc_test.cu
@@ -44,6 +44,32 @@ auto const temp_env = static_cast<cudf::test::TempDirTestEnvironment*>(
     ::testing::AddGlobalTestEnvironment(
         new cudf::test::TempDirTestEnvironment));
 
+template<typename T>
+std::unique_ptr<cudf::experimental::table> create_random_fixed_table(cudf::size_type num_columns, cudf::size_type num_rows, bool include_validity)
+{       
+   auto valids = cudf::test::make_counting_transform_iterator(0, 
+      [](auto i) { 
+        return i % 2 == 0 ? true : false; 
+      }
+    );
+   std::vector<cudf::test::fixed_width_column_wrapper<T>> src_cols(num_columns);
+   for(int idx=0; idx<num_columns; idx++){
+      auto rand_elements = cudf::test::make_counting_transform_iterator(0, [](T i){return rand();});
+      if(include_validity){
+         src_cols[idx] = cudf::test::fixed_width_column_wrapper<T>(rand_elements, rand_elements + num_rows, valids);
+      } else {
+         src_cols[idx] = cudf::test::fixed_width_column_wrapper<T>(rand_elements, rand_elements + num_rows);
+      }
+   }      
+   std::vector<std::unique_ptr<cudf::column>> columns(num_columns);
+   std::transform(src_cols.begin(), src_cols.end(), columns.begin(), [](cudf::test::fixed_width_column_wrapper<T> &in){   
+      auto ret = in.release();
+      ret->has_nulls();
+      return ret;
+   });
+   return std::make_unique<cudf::experimental::table>(std::move(columns));   
+}
+
 // Base test fixture for tests
 struct OrcWriterTest : public cudf::test::BaseFixture {};
 
@@ -69,6 +95,21 @@ using SupportedTimestampTypes = cudf::test::RemoveIf<
     cudf::test::ContainedIn<cudf::test::Types<cudf::timestamp_D>>,
     cudf::test::TimestampTypes>;
 TYPED_TEST_CASE(OrcWriterTimestampTypeTest, SupportedTimestampTypes);
+
+// Base test fixture for chunked writer tests
+struct OrcChunkedWriterTest : public cudf::test::BaseFixture {};
+
+// Typed test fixture for numeric type tests
+template <typename T>
+struct OrcChunkedWriterNumericTypeTest : public OrcChunkedWriterTest {
+  auto type() {
+    return cudf::data_type{cudf::experimental::type_to_id<T>()};
+  }
+};
+
+// Declare typed test cases
+TYPED_TEST_CASE(OrcChunkedWriterNumericTypeTest, cudf::test::NumericTypes);
+
 
 namespace {
 
@@ -389,3 +430,44 @@ TEST_F(OrcWriterTest, HostBuffer) {
   expect_tables_equal(expected->view(), result.tbl->view());
   EXPECT_EQ(expected_metadata.column_names, result.metadata.column_names);
 }
+
+
+TEST_F(OrcChunkedWriterTest, SingleTable)
+{
+  srand(31337);
+  auto table1 = create_random_fixed_table<int>(5, 5, true);      
+
+  auto filepath = temp_env->get_temp_filepath("ChunkedSingle.orc");
+  cudf_io::write_orc_chunked_args args{cudf_io::sink_info{filepath}};
+  auto state = cudf_io::write_orc_chunked_begin(args);  
+  cudf_io::write_orc_chunked(*table1, state);  
+  cudf_io::write_orc_chunked_end(state);    
+
+  cudf_io::read_orc_args read_args{cudf_io::source_info{filepath}};
+  auto result = cudf_io::read_orc(read_args);
+  
+  expect_tables_equal(*result.tbl, *table1);    
+}
+
+TEST_F(OrcChunkedWriterTest, SimpleTable)
+{
+  srand(31337);
+  auto table1 = create_random_fixed_table<int>(5, 5, true);
+  auto table2 = create_random_fixed_table<int>(5, 5, true);
+  
+  auto full_table = cudf::experimental::concatenate({*table1, *table2});          
+
+  auto filepath = temp_env->get_temp_filepath("ChunkedSimple.orc");
+  cudf_io::write_orc_chunked_args args{cudf_io::sink_info{filepath}};
+  auto state = cudf_io::write_orc_chunked_begin(args);  
+  cudf_io::write_orc_chunked(*table1, state);
+  cudf_io::write_orc_chunked(*table2, state);  
+  cudf_io::write_orc_chunked_end(state);    
+
+  cudf_io::read_orc_args read_args{cudf_io::source_info{filepath}};
+  auto result = cudf_io::read_orc(read_args);
+  
+  expect_tables_equal(*result.tbl, *full_table);    
+}
+
+

--- a/cpp/tests/io/orc_test.cu
+++ b/cpp/tests/io/orc_test.cu
@@ -470,4 +470,196 @@ TEST_F(OrcChunkedWriterTest, SimpleTable)
   expect_tables_equal(*result.tbl, *full_table);    
 }
 
+TEST_F(OrcChunkedWriterTest, LargeTables)
+{
+  srand(31337);
+  auto table1 = create_random_fixed_table<int>(512, 4096, true);
+  auto table2 = create_random_fixed_table<int>(512, 8192, true);
+  
+  auto full_table = cudf::experimental::concatenate({*table1, *table2});          
 
+  auto filepath = temp_env->get_temp_filepath("ChunkedLarge.orc");
+  cudf_io::write_orc_chunked_args args{cudf_io::sink_info{filepath}};
+  auto state = cudf_io::write_orc_chunked_begin(args);  
+  cudf_io::write_orc_chunked(*table1, state);
+  cudf_io::write_orc_chunked(*table2, state);  
+  cudf_io::write_orc_chunked_end(state);    
+
+  cudf_io::read_orc_args read_args{cudf_io::source_info{filepath}};
+  auto result = cudf_io::read_orc(read_args);
+  
+  expect_tables_equal(*result.tbl, *full_table);    
+}
+
+TEST_F(OrcChunkedWriterTest, ManyTables)
+{
+  srand(31337);
+  std::vector<std::unique_ptr<table>> tables;
+  std::vector<table_view> table_views;
+  constexpr int num_tables = 96;
+  for(int idx=0; idx<num_tables; idx++){
+    auto tbl = create_random_fixed_table<int>(16, 64, true);
+    table_views.push_back(*tbl);
+    tables.push_back(std::move(tbl));
+  }    
+  
+  auto expected = cudf::experimental::concatenate(table_views);
+
+  auto filepath = temp_env->get_temp_filepath("ChunkedManyTables.orc");
+  cudf_io::write_orc_chunked_args args{cudf_io::sink_info{filepath}};
+  auto state = cudf_io::write_orc_chunked_begin(args); 
+  std::for_each(table_views.begin(), table_views.end(), [&state](table_view const& tbl){
+    cudf_io::write_orc_chunked(tbl, state);
+  });  
+  cudf_io::write_orc_chunked_end(state);    
+
+  cudf_io::read_orc_args read_args{cudf_io::source_info{filepath}};
+  auto result = cudf_io::read_orc(read_args);
+  
+  expect_tables_equal(*result.tbl, *expected);    
+}
+
+TEST_F(OrcChunkedWriterTest, Strings)
+{       
+  std::vector<std::unique_ptr<cudf::column>> cols;
+
+  bool mask1[]    = { 1, 1, 0, 1, 1, 1, 1 };
+  std::vector<const char*> h_strings1 { "four", "score", "and", "seven", "years", "ago", "abcdefgh" };
+  cudf::test::strings_column_wrapper strings1( h_strings1.begin(), h_strings1.end(), mask1 );
+  cols.push_back(strings1.release());  
+  cudf::experimental::table tbl1(std::move(cols));
+
+  bool mask2[]    = { 0, 1, 1, 1, 1, 1, 1 };
+  std::vector<const char*> h_strings2 { "ooooo", "ppppppp", "fff", "j", "cccc", "bbb", "zzzzzzzzzzz" };
+  cudf::test::strings_column_wrapper strings2( h_strings2.begin(), h_strings2.end(), mask2 );
+  cols.push_back(strings2.release());  
+  cudf::experimental::table tbl2(std::move(cols));    
+
+  auto expected = cudf::experimental::concatenate({tbl1, tbl2});  
+
+  auto filepath = temp_env->get_temp_filepath("ChunkedStrings.orc");
+  cudf_io::write_orc_chunked_args args{cudf_io::sink_info{filepath}};
+  auto state = cudf_io::write_orc_chunked_begin(args);   
+  cudf_io::write_orc_chunked(tbl1, state);
+  cudf_io::write_orc_chunked(tbl2, state);
+  cudf_io::write_orc_chunked_end(state);    
+
+  cudf_io::read_orc_args read_args{cudf_io::source_info{filepath}};
+  auto result = cudf_io::read_orc(read_args);      
+
+  expect_tables_equal(*result.tbl, *expected);      
+}
+
+TEST_F(OrcChunkedWriterTest, MismatchedTypes)
+{         
+  srand(31337);
+  auto table1 = create_random_fixed_table<int>(4, 4, true);
+  auto table2 = create_random_fixed_table<float>(4, 4, true);  
+
+  auto filepath = temp_env->get_temp_filepath("ChunkedMismatchedTypes.orc");
+  cudf_io::write_orc_chunked_args args{cudf_io::sink_info{filepath}};
+  auto state = cudf_io::write_orc_chunked_begin(args);  
+  cudf_io::write_orc_chunked(*table1, state);
+  EXPECT_THROW(cudf_io::write_orc_chunked(*table2, state), cudf::logic_error);  
+  cudf_io::write_orc_chunked_end(state);    
+}
+
+TEST_F(OrcChunkedWriterTest, MismatchedStructure)
+{         
+  srand(31337);
+  auto table1 = create_random_fixed_table<int>(4, 4, true);
+  auto table2 = create_random_fixed_table<int>(3, 4, true);  
+
+  auto filepath = temp_env->get_temp_filepath("ChunkedMismatchedStructure.orc");
+  cudf_io::write_orc_chunked_args args{cudf_io::sink_info{filepath}};
+  auto state = cudf_io::write_orc_chunked_begin(args);  
+  cudf_io::write_orc_chunked(*table1, state);
+  EXPECT_THROW(cudf_io::write_orc_chunked(*table2, state), cudf::logic_error);  
+  cudf_io::write_orc_chunked_end(state);    
+}
+
+TYPED_TEST(OrcChunkedWriterNumericTypeTest, UnalignedSize)
+{
+  // write out two 31 row tables and make sure they get
+  // read back with all their validity bits in the right place
+
+  using T = TypeParam;
+
+  int num_els = 31;
+  std::vector<std::unique_ptr<cudf::column>> cols;
+
+  bool mask[]    = { 0, 1, 1, 1, 1, 1, 1, 1,  1, 1, 1, 1, 1, 1, 1, 1,  1, 1, 1, 1, 1, 1, 1, 1,  1, 1, 1, 1, 1, 1, 1 };
+
+  T c1a[]        = { 5, 5, 5, 5, 5, 5, 5, 5,  5, 5, 5, 5, 5, 5, 5, 5,  5, 5, 5, 5, 5, 5, 5, 5,  5, 5, 5, 5, 5, 5, 5 };
+  T c1b[]        = { 6, 6, 6, 6, 6, 6, 6, 6,  6, 6, 6, 6, 6, 6, 6, 6,  6, 6, 6, 6, 6, 6, 6, 6,  6, 6, 6, 6, 6, 6, 6 };
+  column_wrapper<T> c1a_w(c1a, c1a + num_els, mask);
+  column_wrapper<T> c1b_w(c1b, c1b + num_els, mask);
+  cols.push_back(c1a_w.release());
+  cols.push_back(c1b_w.release());
+  cudf::experimental::table tbl1(std::move(cols));
+
+  T c2a[]        = { 8, 8, 8, 8, 8, 8, 8, 8,  8, 8, 8, 8, 8, 8, 8, 8,  8, 8, 8, 8, 8, 8, 8, 8,  8, 8, 8, 8, 8, 8, 8 };
+  T c2b[]        = { 9, 9, 9, 9, 9, 9, 9, 9,  9, 9, 9, 9, 9, 9, 9, 9,  9, 9, 9, 9, 9, 9, 9, 9,  9, 9, 9, 9, 9, 9, 9 };
+  column_wrapper<T> c2a_w(c2a, c2a + num_els, mask);
+  column_wrapper<T> c2b_w(c2b, c2b + num_els, mask);
+  cols.push_back(c2a_w.release());
+  cols.push_back(c2b_w.release());
+  cudf::experimental::table tbl2(std::move(cols));
+
+  auto expected = cudf::experimental::concatenate({tbl1, tbl2});  
+
+  auto filepath = temp_env->get_temp_filepath("ChunkedUnalignedSize.orc");
+  cudf_io::write_orc_chunked_args args{cudf_io::sink_info{filepath}};
+  auto state = cudf_io::write_orc_chunked_begin(args);   
+  cudf_io::write_orc_chunked(tbl1, state);
+  cudf_io::write_orc_chunked(tbl2, state);
+  cudf_io::write_orc_chunked_end(state);    
+
+  cudf_io::read_orc_args read_args{cudf_io::source_info{filepath}};
+  auto result = cudf_io::read_orc(read_args);      
+
+  expect_tables_equal(*result.tbl, *expected);      
+}
+
+TYPED_TEST(OrcChunkedWriterNumericTypeTest, UnalignedSize2)
+{
+  // write out two 33 row tables and make sure they get
+  // read back with all their validity bits in the right place
+
+  using T = TypeParam;
+
+  int num_els = 33;
+  std::vector<std::unique_ptr<cudf::column>> cols;
+
+  bool mask[]    = { 0, 1, 1, 1, 1, 1, 1, 1,  1, 1, 1, 1, 1, 1, 1, 1,  1, 1, 1, 1, 1, 1, 1, 1,  1, 1, 1, 1, 1, 1, 1, 1, 1 };
+
+  T c1a[]        = { 5, 5, 5, 5, 5, 5, 5, 5,  5, 5, 5, 5, 5, 5, 5, 5,  5, 5, 5, 5, 5, 5, 5, 5,  5, 5, 5, 5, 5, 5, 5, 5, 5 };
+  T c1b[]        = { 6, 6, 6, 6, 6, 6, 6, 6,  6, 6, 6, 6, 6, 6, 6, 6,  6, 6, 6, 6, 6, 6, 6, 6,  6, 6, 6, 6, 6, 6, 6, 6, 6 };
+  column_wrapper<T> c1a_w(c1a, c1a + num_els, mask);
+  column_wrapper<T> c1b_w(c1b, c1b + num_els, mask);
+  cols.push_back(c1a_w.release());
+  cols.push_back(c1b_w.release());
+  cudf::experimental::table tbl1(std::move(cols));
+
+  T c2a[]        = { 8, 8, 8, 8, 8, 8, 8, 8,  8, 8, 8, 8, 8, 8, 8, 8,  8, 8, 8, 8, 8, 8, 8, 8,  8, 8, 8, 8, 8, 8, 8, 8, 8 };
+  T c2b[]        = { 9, 9, 9, 9, 9, 9, 9, 9,  9, 9, 9, 9, 9, 9, 9, 9,  9, 9, 9, 9, 9, 9, 9, 9,  9, 9, 9, 9, 9, 9, 9, 9, 9 };
+  column_wrapper<T> c2a_w(c2a, c2a + num_els, mask);
+  column_wrapper<T> c2b_w(c2b, c2b + num_els, mask);
+  cols.push_back(c2a_w.release());
+  cols.push_back(c2b_w.release());
+  cudf::experimental::table tbl2(std::move(cols));
+
+  auto expected = cudf::experimental::concatenate({tbl1, tbl2});  
+
+  auto filepath = temp_env->get_temp_filepath("ChunkedUnalignedSize2.orc");
+  cudf_io::write_orc_chunked_args args{cudf_io::sink_info{filepath}};
+  auto state = cudf_io::write_orc_chunked_begin(args);   
+  cudf_io::write_orc_chunked(tbl1, state);
+  cudf_io::write_orc_chunked(tbl2, state);
+  cudf_io::write_orc_chunked_end(state);    
+
+  cudf_io::read_orc_args read_args{cudf_io::source_info{filepath}};
+  auto result = cudf_io::read_orc(read_args);      
+
+  expect_tables_equal(*result.tbl, *expected);      
+}

